README
Kudzu FAQ:
----------

1) So, what is kudzu?

kudzu is the hardware autodetection & configuration tool, originally
introduced in Red Hat Linux 6.1. It detects changes in the system's
hardware configuration, and gives the option of configuring
added or removed devices. It's set up by default to run at
each boot.

2) What type of devices does kudzu detect?

Currently, kudzu detects:

- PCI cards
- SBUS cards
- serial devices (PnP devices, mice, modems)
- parallel port devices (printers)
- SCSI devices (provided the appropriate SCSI modules are loaded)
- IDE devices
- PS/2 mice
- Keyboards on SPARC

More devices and types will be added in the future.

3) What is kudzu doing when it says 'Configuring device X...'?

Configuration:
- Network:
	- adds an 'ethX' alias, if necessary
	- migrates old device config, or creates a new one.
- SCSI:
	- adds a 'scsi_hostadapter' alias
- Video card:
	- runs Xconfigurator
- Sound card:
	- runs sndconfig
- Mouse
	- makes /dev/mouse link
	- runs mouseconfig
- Modem
	- makes /dev/modem link
- CD-ROM
        - makes /dev/cdrom link
- Scanner
	- makes /dev/scanner link
- Keyboard
	- runs kbdconfig
	- for serial console makes sure /etc/inittab and /etc/securetty
	  are ready for serial console operation

Un-configuration:
- Network:
	- removes ethX alias
- SCSI:
	- removes scsi_hostadapterX alias
- Mouse:
	- removes /dev/mouse link
- Modem:
	- removes /dev/modem link
- CD-ROM:
	- removes /dev/cdrom link
- Scanner:
	- removes /dev/scanner link

4) Where do I submit bug reports about kudzu?

http://developer.redhat.com/bugzilla/

5) Are there any known bugs in kudzu?

The serial probe can confuse devices if they are
currently in use. (modems, mice) The PS/2 probe
also will cause the mouse to act strangely under X.

Running the serial probe on a machine that is acting
as a serial console can produce odd results.



debian/changelog
kudzu-kanotix (1.2.34.3-2) ORPHANED; urgency=low

  * NOT RELEASED YET
  * ORPHANED
  * only use python-2.4.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 11 Apr 2006 20:20:31 +0200

kudzu-kanotix (1.2.34.3-1) unstable; urgency=low

  * New upstream release:
    - fix scsi probe to not return devices of the wrong class
    - special-case xen to not use vm86 (works around #179013)
    - switch at runtime between vm86 and x86emu on i386. Fixes vbe/ddc on
      Xen and i386-on-x86_64
    - silence some error messages
    - fix i2o again (#182522)
    - fix potential segfault with USB network devices (#182006)
    - further x86emu-related fixes (#181467)
  * don't conflict with kudzu-vesa anymore, as X.org 7 depends on xresprobe
    which in turn provides kudzu-vesa.
  * add plain python to dependencies.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 11 Apr 2006 19:46:04 +0200

kudzu-kanotix (1.2.30-3) unstable; urgency=low

  * fix dependencies.
  * ensure that kudzu-vesa is removed.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 18 Feb 2006 00:05:55 +0100

kudzu-kanotix (1.2.30-2) unstable; urgency=low

  * ddcprobe, ddcxinfo, modetest and svgamodes are gone, remove
    kudzu-kanotix-vesa as it is no longer used.
  * re- create packaging from scratch.
  * remove dietlibc-dev from build-depends, as it is no longer used by
    upstream.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Fri, 17 Feb 2006 23:42:41 +0100

kudzu-kanotix (1.2.30-1) unstable; urgency=low

  * New upstream release:
    - revert changes for video probing classes - some correct devices are
      0380
    - silence!
    - vbe/edid probing on x86_64 via x86emu - thanks to Matthew Garrett
      & Tollef Fog Heen
    - restrict video devices to PCI class 0300, correctly (should solve #176978)
    - restrict video devices to PCI class 0300 (should solve #176978)
    - translation updates
    - in installer environment, sg isn't loaded, so there are deviceless
      scsi devs (#180378)
    - fix xenblk/i2o sysfs probing
    - fix nvidia segfault
    - fix xen device probing when modules aren't loaded
    - more forcedeth shenaningans (#178557)
    - switch scsi probe over to sysfs
    - fix potential segfault in DDC probing code
    - don't configure (or unconfigure) storage adapters (#177740)
    - add missing fchdir in pcmcia code
    - fix segfault if pcmcia network devices are found before their drivers
      are loaded (#174341)
    - handle pcilib string returns (#176490, #176724)
  * drop makefile.dpatch, it is no longer necessary.
  * bump compat level to 5.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Fri, 17 Feb 2006 20:22:42 +0100

kudzu-kanotix (1.2.16-1) unstable; urgency=low

  * new upstream release:
    - revert LRMI changes (#175677)
    - use ACPI EDID info if available
    - add support for mambo virtual devices (#173307, <dwmw2@redhat.com>)
    - integrate patch, bump version
    - Hopefully fix ddcprobe to work with > 2.6.14 kernels.
    - allow multiple files for video aliases under /usr/share/hwdata/videoaliases
    - fix xen driver names
    - add some basic xen device probing
  * remove virtual knoppix packages.
  * bump compat level to 5.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 15 Dec 2005 11:52:37 +0100

kudzu-kanotix (1.2.9-1) unstable; urgency=low

  * new upstream version:
    - move kudzu to /sbin (since we no longer use newt (#74736))
    - don't congfigure usb/firewire controllers, modems, scanners
       in kudzu program (as such configurations aren't used)
    - fix crash in sortNetDevices (#169003)
    - fix fbProbe to work with X drivers, not card entries
    - fix crash in matchNetDevices (#168689)
    - fix module_upgrade
    - remove obsolete updfstab code
    - ABI change: drivers are no longer set to unknown/ignore/disabled;
      they are just left as NULL
    - remove support for loading modules; it's not used by any library
      consumers
    - remove support for system-config-mouse, as it's no longer shipped
    - read hwaddrs for network devices from sysfs, not ethtool (and
      conflict with older kernels that don't support that)
    - port pcmcia probe to new model
    - fix usb device list double-free
    - fix passing of filenames to (pci|usb)ReadDrivers()
    - switch pci, usb probing to use modules.alias
    - switch usb probe to use sysfs
    - remove pcitable support
    - X drivers are now the video.xdriver field of CLASS_VIDEO
      (framebuffer drivers will be returned if they match)

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sun,  2 Oct 2005 16:26:09 +0200

kudzu-kanotix (1.1.122-1) unstable; urgency=low

  * new upstream release:
    - don't rely on pcimap for 8139too/8139cp; hardcode the logic (#157783)
    - make sure kernel version is always initialized (fixes python bindings)
  * adapt Makefile for gcc-4.0.2 woes.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sun,  2 Oct 2005 15:59:42 +0200

kudzu-kanotix (1.1.120-1) unstable; urgency=low

  * New upstream release:
    - fix macio overzealous snd-powermac probe (#166011, <dwmw2@redhat.com>)
    - fix overriding of kernel version
      - allow using that in module_upgrade
    - remove usb, firewire special cases in pci probing
  * rediff build-fix.
  * fix changelog.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 23 Aug 2005 13:59:39 +0200

kudzu-kanotix (1.1.119-1) unstable; urgency=low

  * New upstream release:
    - make sure changing network devices are properly caught (#141338)
  * bump standards version to 3.6.2.
  * adapt build-depends to debhelper 4.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 30 Jul 2005 16:11:12 +0200

kudzu-kanotix (1.1.118-1) unstable; urgency=low

  * new upstream version:
    - make sure ISAPNP devices don't end up as CLASS_UNSPEC (#163949)
    - fix usb - class/subclass/protocol are hex, not decimal (#160450)

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Mon, 25 Jul 2005 16:25:30 +0200

kudzu-kanotix (1.1.116.2-2) unstable; urgency=low


pci.c
/* Copyright 1999-2005 Red Hat, Inc.
 *
 * This software may be freely redistributed under the terms of the GNU
 * public license.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <ctype.h>
#include <dirent.h>
#include <fcntl.h>
#include <setjmp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <libgen.h>

#include <sys/stat.h>
#include <sys/types.h>

#include <pci/pci.h>

#include "pci.h"
#include "pciserial.h"

#include "kudzuint.h"

static struct pci_access *pacc=NULL;
static jmp_buf pcibuf;
static char *pcifiledir = NULL;
static void pciWriteDevice(FILE *file, struct pciDevice *dev);

static int devCmp2(const void * a, const void * b) {
    const struct pciDevice * one = a;
    const struct pciDevice * two = b;
    int x=0,y=0,z=0,xx=0,yy=0;
    
    x = (one->vendorId - two->vendorId);
    xx = (one->subVendorId - two->subVendorId);
    y = (one->deviceId - two->deviceId);
    yy = (one->subDeviceId - two->subDeviceId);
    if (one->pciType && two->pciType)
	  z = (one->pciType - two->pciType);
    if (x)
      return x;
    else if (y)
      return y;
    else if (one->subVendorId != 0xffff && two->subVendorId != 0xffff) {
	    if (xx)
	      return xx;
	    else if (yy)
	      return yy;
    } 
	return z;
}

static void pciFreeDevice(struct pciDevice *dev) {
    freeDevice((struct device *)dev);
}

static void pciWriteDevice(FILE *file, struct pciDevice *dev) {
	writeDevice(file, (struct device *)dev);
	fprintf(file,"vendorId: %04x\ndeviceId: %04x\nsubVendorId: %04x\nsubDeviceId: %04x\npciType: %d\npcidom: %4x\npcibus: %2x\npcidev: %2x\npcifn: %2x\n",
		dev->vendorId,dev->deviceId,dev->subVendorId,dev->subDeviceId,dev->pciType,dev->pcidom,dev->pcibus,dev->pcidev,dev->pcifn);
}

static int pciCompareDevice(struct pciDevice *dev1, struct pciDevice *dev2)
{
	int x,y;
	
	x = compareDevice((struct device *)dev1,(struct device *)dev2);
	if (x && x!=2) 
	  return x;
	if ((y=devCmp2( (void *)dev1, (void *)dev2 )))
	  return y;
	return x;
}
			    
struct pciDevice * pciNewDevice(struct pciDevice *dev) {
    struct pciDevice *ret;
    
    ret = malloc(sizeof(struct pciDevice));
    memset(ret,'\0',sizeof(struct pciDevice));
    ret=(struct pciDevice *)newDevice((struct device *)dev,(struct device *)ret);
    ret->bus = BUS_PCI;
    /* Older entries will come in with subvendor & subdevice IDs of 0x0.
     * 0x0 is a *valid* entry, though, so set it to something invalid. */
    ret->subVendorId = 0xffff;
    if (dev && dev->bus == BUS_PCI) {
	ret->vendorId = dev->vendorId;
	ret->deviceId = dev->deviceId;
	ret->pciType = dev->pciType;
	ret->subVendorId = dev->subVendorId;
        ret->subDeviceId = dev->subDeviceId;
	ret->pcidom = dev->pcidom;
	ret->pcibus = dev->pcibus;
	ret->pcidev = dev->pcidev;
	ret->pcifn = dev->pcifn;
    } else {
	ret->pciType = PCI_UNKNOWN;
    }
    ret->newDevice = pciNewDevice;
    ret->freeDevice = pciFreeDevice;
    ret->writeDevice = pciWriteDevice;
    ret->compareDevice = pciCompareDevice;
    return ret;
}

static unsigned int kudzuToPci(enum deviceClass class) {
    switch (class) {
     case CLASS_UNSPEC:
	return 0;
     case CLASS_OTHER:
	return 0;
     case CLASS_NETWORK:
	return PCI_BASE_CLASS_NETWORK;
     case CLASS_VIDEO:
	return PCI_BASE_CLASS_DISPLAY;
     case CLASS_AUDIO:
	return PCI_CLASS_MULTIMEDIA_AUDIO;
     case CLASS_SCSI:
	return PCI_CLASS_STORAGE_SCSI;
     case CLASS_FLOPPY:
	return PCI_CLASS_STORAGE_FLOPPY;
     case CLASS_RAID:
	return PCI_CLASS_STORAGE_RAID;
     case CLASS_CAPTURE:
	return PCI_CLASS_MULTIMEDIA_VIDEO;
     case CLASS_MODEM:
	return PCI_CLASS_COMMUNICATION_SERIAL;
     case CLASS_MOUSE: /* !?!? */
	return PCI_CLASS_INPUT_MOUSE;
     case CLASS_USB:
	return PCI_CLASS_SERIAL_USB;
     case CLASS_FIREWIRE:
	return PCI_CLASS_SERIAL_FIREWIRE;
     case CLASS_SOCKET:
	return PCI_CLASS_BRIDGE_CARDBUS;
     case CLASS_IDE:
	return PCI_CLASS_STORAGE_IDE;
     default:
	return 0;
    }
}

static enum deviceClass pciToKudzu(unsigned int class) {
    
    if (!class) return CLASS_UNSPEC;
    switch (class >> 8) {
     case PCI_BASE_CLASS_DISPLAY:
	return CLASS_VIDEO;
    case PCI_BASE_CLASS_NETWORK:
	return CLASS_NETWORK;
     default:
	break;
    }
    switch (class) {
     case PCI_CLASS_STORAGE_SCSI:
     case PCI_CLASS_SERIAL_FIBER:
	return CLASS_SCSI;
     case PCI_CLASS_STORAGE_FLOPPY:
	return CLASS_FLOPPY;
     case PCI_CLASS_STORAGE_RAID:
	return CLASS_RAID;
     case PCI_CLASS_MULTIMEDIA_AUDIO:
     /* HD Audio */
     case 0x0403:
	return CLASS_AUDIO;
     case PCI_CLASS_INPUT_MOUSE:
	return CLASS_MOUSE;
     case PCI_CLASS_MULTIMEDIA_OTHER:
     case PCI_CLASS_MULTIMEDIA_VIDEO:
	return CLASS_CAPTURE;
     case PCI_CLASS_COMMUNICATION_SERIAL:
     case PCI_CLASS_COMMUNICATION_OTHER:
	return CLASS_MODEM;
     case PCI_CLASS_NOT_DEFINED_VGA:
	return CLASS_VIDEO;
     /* Fix for one of the megaraid variants.
      * It claims to be an I2O controller. */
     case 0x0e00:
	return CLASS_SCSI;
     case PCI_CLASS_SERIAL_USB:
	return CLASS_USB;
     case PCI_CLASS_SERIAL_FIREWIRE:
	return CLASS_FIREWIRE;
     case PCI_CLASS_BRIDGE_CARDBUS:
	return CLASS_SOCKET;
     case PCI_CLASS_NETWORK_ETHERNET:
	return CLASS_NETWORK;
     case PCI_CLASS_NETWORK_TOKEN_RING:
	return CLASS_NETWORK;
     case PCI_CLASS_NETWORK_FDDI:
	return CLASS_NETWORK;
     case PCI_CLASS_NETWORK_ATM:
	return CLASS_NETWORK;
     case PCI_CLASS_STORAGE_IDE:
	return CLASS_IDE;
     default:
	return CLASS_OTHER;
    }
}

static void readVideoAliasesDir(char *dirname) {
	DIR *d;
	struct dirent *entry;
	
	d = opendir(dirname);
	if (!d) return;
	
	while ((entry = readdir(d))) {
		char *path;
		
		if (entry->d_name[0] == '.') continue;
		
		asprintf(&path,"%s/%s",dirname,entry->d_name);
		aliases = readAliases(aliases, path, "pcivideo");
		free(path);
	}
	closedir(d);
}

int pciReadDrivers(char *filename) {
	char *p;
	struct stat sbuf;
	
	aliases = readAliases(aliases, filename, "pci");

	if (filename) {
		pcifiledir = dirname(strdup(filename));
		asprintf(&p,"%s/videoaliases",pcifiledir);
	        if (!stat(p,&sbuf)) {
			return 0;
		}
		if (S_ISDIR(sbuf.st_mode)) {
			readVideoAliasesDir(p);
		} else
			aliases = readAliases(aliases, p, "pcivideo");
		free(p);
		return 0;
	} else {
		int x;
		struct stat sbuf;
		char *paths[] = { "/usr/share/hwdata/videoaliases", "/etc/videoaliases",
				"/modules/videoaliases", "./videoaliases" , NULL };
		
		for (x = 0; paths[x] ; x++) {
			if (!stat(paths[x],&sbuf)) {
				p = paths[x];
				break;
			}
		}
		if (!paths[x])
			return 0;
		if (S_ISDIR(sbuf.st_mode)) {
			readVideoAliasesDir(p);
		} else 
			aliases = readAliases(aliases, p, "pcivideo");
	}
	return 0;
}

void pciFreeDrivers(void) {
}

static void pcinull(char * foo, ...)
{
}

static void pcibail(char * foo, ...)
{
    longjmp(pcibuf,1);
}

static int isDisabled(struct pci_dev *p, u_int8_t config[256]) {
	int disabled;
	int i;
#ifdef __i386__
	int limit = 6;
#else
	int limit = 1;
#endif	
	unsigned int devtype, command;
	
	devtype = p->device_class;
	if (p->irq || pciToKudzu(devtype) != CLASS_VIDEO) {
		return 0;
	}
	/* Only check video cards for now. */
	command = config[PCI_COMMAND];
	disabled = 0;
	for (i = 0; i < limit ; i++) {
		int x = PCI_BASE_ADDRESS_0 + 4 * i;
		pciaddr_t pos = p->base_addr[i];
		pciaddr_t len = (p->known_fields & PCI_FILL_SIZES) ? p->size[i] : 0;
		u32 flag = config[x+3] << 24 | config[x+2] << 16 | config[x+1] << 8 | config[x];
		if ((flag == 0xffffffff || !flag) && !pos && !len)
		  continue;
		disabled++;
		if ((flag & PCI_BASE_ADDRESS_SPACE_IO) && (command & PCI_COMMAND_IO))
		  disabled--;
		else if (command & PCI_COMMAND_MEMORY)
		  disabled--;			
	}
	return disabled;
}

static char *getNetworkPath(struct pciDevice *dev) {
	char *path, *ret = NULL;
	DIR *d;
	struct dirent *entry;
	
	asprintf(&path,"/sys/bus/pci/devices/%04x:%02x:%02x.%x",dev->pcidom,
		 dev->pcibus, dev->pcidev, dev->pcifn);
	d = opendir(path);
	if (!d) {
		free(path);
		return NULL;
	}
	while ((entry = readdir(d))) {
		if (!strncmp(entry->d_name,"net:",4)) {
			if (ret) free(ret);
			asprintf(&ret,"%s/%s",path,entry->d_name);
		}
	}
	free(path);
	return ret;
}

struct device * pciProbe(enum deviceClass probeClass, int probeFlags, struct device *devlist) {
    struct pci_dev *p;
    /* This should be plenty. */
    int cardbus_bridges[32];
    int bridgenum = 0;
    int init_list = 0;
    char pcifile[128];
    unsigned int type = kudzuToPci(probeClass),devtype;
    
    if (
	(probeClass & CLASS_OTHER) ||
	(probeClass & CLASS_NETWORK) ||
	(probeClass & CLASS_SCSI) ||
	(probeClass & CLASS_IDE) ||
	(probeClass & CLASS_VIDEO) ||
	(probeClass & CLASS_AUDIO) ||
	(probeClass & CLASS_MODEM) ||
	(probeClass & CLASS_USB) ||
	(probeClass & CLASS_FIREWIRE) ||
	(probeClass & CLASS_SOCKET) ||
	(probeClass & CLASS_CAPTURE) ||
	(probeClass & CLASS_RAID)) {
	pacc = pci_alloc();
	if (!pacc) return devlist;
	if (!getAliases(aliases, "pci")) {
		pciReadDrivers(NULL);
		init_list = 1;
	}
	pacc->debug=pacc->warning=pcinull;
	pacc->error=pcibail;
	if (!access("/usr/share/hwdata/pci.ids", R_OK))
		    pacc->id_file_name = "/usr/share/hwdata/pci.ids";
	else if (!access("/etc/pci.ids", R_OK))
		    pacc->id_file_name = "/etc/pci.ids";
	else if (!access("/modules/pci.ids", R_OK))
		    pacc->id_file_name = "/modules/pci.ids";
	else if (!access("./pci.ids", R_OK))
		    pacc->id_file_name = "./pci.ids";
	else if (pcifiledir) {
		snprintf(pcifile,128,"%s/pci.ids",pcifiledir);
		if (!access(pcifile, R_OK))
			pacc->id_file_name = pcifile;
	}
		
		    
	if (!setjmp(pcibuf)) {
	    int order=0;
		
	    pci_init(pacc);
	    pci_scan_bus(pacc);
		
	    memset(cardbus_bridges,'\0',32);
	    /* enumerate cardbus bridges first */
	    for (p = pacc->devices; p; p=p->next) {
		u_int8_t config[256];
		
		memset(config,'\0',256);
	        pci_read_block(p, 0, config, 64);
		if ((config[PCI_HEADER_TYPE] & 0x7f) == PCI_HEADER_TYPE_CARDBUS) {
		    /* Cardbus bridge */
		    pci_read_block(p, 64, config+64, 64);
		    for (bridgenum=0; cardbus_bridges[bridgenum];bridgenum++);
		    cardbus_bridges[bridgenum] = config[PCI_CB_CARD_BUS];
		}
	    }
	    
	    for (p = pacc->devices; p; p=p->next) {
		char *t;
		char *drv, *x_drv;
		u_int8_t config[256];
		int bustype;
		unsigned int subvend, subdev;
		struct pciDevice *dev,*a_dev;
		
		memset(config,'\0',256);
		pci_read_block(p, 0, config, 64);
		if ((config[PCI_HEADER_TYPE] & 0x7f) == PCI_HEADER_TYPE_CARDBUS) {
		    /* Cardbus bridge */
		    pci_read_block(p, 64, config+64, 64);
		    subvend = config[PCI_CB_SUBSYSTEM_VENDOR_ID+1] << 8 | config[PCI_CB_SUBSYSTEM_VENDOR_ID];
		    subdev = config[PCI_CB_SUBSYSTEM_ID+1] << 8 | config[PCI_CB_SUBSYSTEM_ID];
		} else {
		    subvend = config[PCI_SUBSYSTEM_VENDOR_ID+1] << 8 | config[PCI_SUBSYSTEM_VENDOR_ID];
		    subdev = config[PCI_SUBSYSTEM_ID+1] << 8 | config[PCI_SUBSYSTEM_ID];
		}
		pci_fill_info(p, PCI_FILL_IDENT | PCI_FILL_CLASS | PCI_FILL_IRQ | PCI_FILL_BASES | PCI_FILL_ROM_BASE | PCI_FILL_SIZES);
		bustype = PCI_NORMAL;
	        for (bridgenum=0; cardbus_bridges[bridgenum]; bridgenum++) {
			if (p->bus == cardbus_bridges[bridgenum])
			  bustype = PCI_CARDBUS;
		}
		dev = pciNewDevice(NULL);
		dev->vendorId = p->vendor_id;
		dev->deviceId = p->device_id;
		dev->subVendorId = subvend;
		dev->subDeviceId = subdev;
		dev->pciType = bustype;
		asprintf(&t,"v%08Xd%08Xsv%08Xsd%08Xbc%02Xsc%02Xi%02x",p->vendor_id,p->device_id,
			 subvend, subdev,(u_int8_t)(p->device_class >> 8),(u_int8_t)(p->device_class),config[PCI_CLASS_PROG]);
		drv = aliasSearch(aliases, "pci", t);
		x_drv = aliasSearch(aliases, "pcivideo", t);
		free(t);
		if (drv)
			dev->driver = strdup(drv);
		devtype = p->device_class;
		if (x_drv) {
			dev->classprivate = strdup(x_drv);
		}
		/* Check for an i2o device. Note that symbios controllers
		 * also need i2o_scsi module. Dunno how to delineate that
		 * here. */
		if (devtype == 0x0e00) {
                    if ((config[PCI_CLASS_PROG] == 0 || config[PCI_CLASS_PROG] == 1)) { 
				if (dev->driver) free(dev->driver);
				dev->driver = strdup("i2o_block");
			}
		}
		if (devtype == PCI_CLASS_BRIDGE_CARDBUS) {
			if (dev->driver) free(dev->driver);
			dev->driver = strdup("yenta_socket");
		}
		if (isDisabled(p, config)) {
			if (dev->driver) free(dev->driver);
			dev->driver = NULL;
		}
	        if (dev->pciType == PCI_CARDBUS) {
			dev->detached = 1;
		}
		/* Sure, reuse a pci id for an incompatible card. */
		if (dev->vendorId == 0x10ec && dev->deviceId == 0x8139) {
			if (config[PCI_REVISION_ID] >= 0x20) {
				if (dev->driver) free(dev->driver);
				dev->driver = strdup("8139cp");
			} else {
				if (dev->driver) free(dev->driver);
				dev->driver = strdup("8139too");
			}
		}
		/* nForce4 boards show up with their ethernet controller
		 * as a bridge; hack it */
		if (dev->vendorId == 0x10de && (dev->driver != NULL) 
                    && !strcmp(dev->driver,"forcedeth")
		    && devtype == PCI_CLASS_BRIDGE_OTHER)
			    devtype = PCI_CLASS_NETWORK_ETHERNET;
		dev->pcidom = p->domain;
		dev->pcibus = p->bus;
		dev->pcidev = p->dev;
		dev->pcifn = p->func;
		while (1) {
			static int size = 128;
			
			if (dev->desc) free(dev->desc);
			dev->desc = malloc(size);
			dev->desc = pci_lookup_name(pacc, dev->desc, size,
					    PCI_LOOKUP_VENDOR | PCI_LOOKUP_DEVICE,
					    p->vendor_id, p->device_id, 0, 0);
			if (!strncmp(dev->desc,"<pci_lookup_name:",17)) {
				size *= 2;
				dev->desc = NULL;
			} else {
				break;
			}
		}
		    
		if ( (probeFlags & PROBE_ALL) || (dev->driver)) {
		    if (!type || (type<0xff && (type==devtype>>8))
			|| (type== kudzuToPci (pciToKudzu (devtype)))) {
			a_dev = pciNewDevice(dev);
			a_dev->type = pciToKudzu(devtype);
			if (x_drv) {
				a_dev->classprivate = strdup(x_drv);
				a_dev->type = CLASS_VIDEO;
			}
			switch (a_dev->type) {
			case CLASS_NETWORK:
				{
				char *netpath;
				
				netpath = getNetworkPath(dev);
				
				if (!netpath || __getNetworkDevAndAddr((struct device *)a_dev,netpath)) {
					if (devtype == PCI_CLASS_NETWORK_TOKEN_RING)
						a_dev->device = strdup("tr");
					else if (devtype == PCI_CLASS_NETWORK_FDDI)
						a_dev->device = strdup("fddi");
					else
						a_dev->device = strdup("eth");
				}
				if (netpath) free(netpath);
				break;
				}
			case CLASS_MODEM:
				checkPCISerial (a_dev, p);
				break;
			default:
				break;
			}
			a_dev->index = order;
			order++;
			if (devlist) {
			    a_dev->next = devlist;
			}
			devlist = (struct device *)a_dev;
		    } 
		}
		pciFreeDevice(dev);
	    }
	    pci_cleanup(pacc);
	}
    }
    if (init_list)
	  pciFreeDrivers();
    return devlist;
}

